### PR TITLE
Let DynamicDataOptions.Default specify case-insensitive serialization options

### PR DESCRIPTION
**Pull request by annelo-msft**
DynamicData doesn't currently allow a cast to a POCO if the JSON member name casing doesn't match the POCO property name casing.  This PR enables the following to succeed:

```csharp
    dynamic data = BinaryData.FromString(
        """
        {
            "message": "Hi",
            "number" : 5
        }
        """).ToDynamicFromJson(DynamicDataOptions.Default);
    SampleModel model = (SampleModel)data;
```

where SampleModel is defined as:
```csharp
    public class SampleModel : IEquatable<SampleModel>
    {
        public SampleModel() { }

        public string Message { get; set; }
        public int Number { get; set; }
     }
```

### CODE REVIEW

File: sdk/core/Azure.Core/src/DynamicData/DynamicData.cs [hunk 1]
```diff
@@ -31,11 +31,15 @@ public sealed partial class DynamicData : IDisposable
...
         private DynamicDataOptions _options;
+        private JsonSerializerOptions _serializerOptions;
 
         internal DynamicData(MutableJsonElement element, DynamicDataOptions options = default)
         {
             _element = element;
             _options = options;
+            _serializerOptions = options.NameMapping == DynamicDataNameMapping.None ?
+                new JsonSerializerOptions() :
+                new JsonSerializerOptions() { PropertyNameCaseInsensitive = true };
```
Comment (heaths) [inline sdk/core/Azure.Core/src/DynamicData/DynamicData.cs:42]: Given the other two current options, this doesn't feel too "loose"? Granted, I don't know how else you'd qualify this in the enum. Probably fine. Just wondering.

**Review by heaths — APPROVED** [review]

**Comment by heaths** [comment]
Generally, should we use our own `ObjectSerializer` that could support other formats? Though, not sure we're supposed to really rely on that now.

File: sdk/core/Azure.Core/src/DynamicData/DynamicData.cs (hunk 1 above)
Comment (annelo-msft) [reply sdk/core/Azure.Core/src/DynamicData/DynamicData.cs]: I changed it to use PropertyNamingPolicy instead -- this should more closely match the DynamicDataOptions value.

File: sdk/core/Azure.Core/src/DynamicData/DynamicData.cs [hunk 2]
```diff
@@ -30,16 +30,21 @@ public sealed partial class DynamicData : IDisposable
...
             _options = options;
-            _serializerOptions = options.NameMapping == DynamicDataNameMapping.None ?
-                new JsonSerializerOptions() :
-                new JsonSerializerOptions() { PropertyNameCaseInsensitive = true };
+
+            _serializerOptions = new JsonSerializerOptions();
+
+            if (options.NameMapping == DynamicDataNameMapping.PascalCaseGettersCamelCaseSetters)
+            {
+                _serializerOptions.PropertyNamingPolicy = JsonNamingPolicy.CamelCase;
```
[the remark this answers is not in the document]
Comment (heaths) [reply sdk/core/Azure.Core/src/DynamicData/DynamicData.cs:46]: If you don't feel that subtlety is worth the effort, that's totally okay. I was just pondering the ramifications of being loose with the deserialization when the other members besides `None` are quite specific.

**Comment by annelo-msft** [comment]
Closing as design approach has changed.

**Pull request closed without merging**